### ISSUE TITLE
Better padding API

### ISSUE THREAD

**Issue by jbschlosser** [pytorch/pytorch#153854]
Users often complain that the API of `torch.nn.functional.pad()` is awkward to use or incomplete. We should provide a better alternative (e.g. a new `torch.pad()`).

**Current API of `torch.nn.functional.pad()`:**
```python
t4d = torch.empty(3, 3, 4, 2)
p1d = (1, 1) # pad last dim by 1 on each side
out = F.pad(t4d, p1d, "constant", 0)  # effectively zero padding
print(out.size())  # torch.Size([3, 3, 4, 4])
p2d = (1, 1, 2, 2) # pad last dim by (1, 1) and 2nd to last by (2, 2)
out = F.pad(t4d, p2d, "constant", 0)
print(out.size())  # torch.Size([3, 3, 8, 4])
t4d = torch.empty(3, 3, 4, 2)
p3d = (0, 1, 2, 1, 3, 3) # pad by (0, 1), (2, 1), and (3, 3)
out = F.pad(t4d, p3d, "constant", 0)
print(out.size())  # torch.Size([3, 9, 7, 3])
```

Padding input is a single tuple of even length specifying (before, after) padding for some number of dimensions, starting with the last dimension and working back. Padding does not have to be specified for all dimensions.

**Proposed torch.pad() API options:**
1. Match numpy's [`pad()`](https://numpy.org/doc/stable/reference/generated/numpy.pad.html) API (see #60294 / #https://github.com/pytorch/pytorch/pull/61635).

```python
import numpy as np
a = [1, 2, 3, 4, 5]
np.pad(a, (2, 3), 'constant', constant_values=(4, 6))
array([4, 4, 1, ..., 6, 6, 6])
np.pad(a, (2, 3), 'edge')
array([1, 1, 1, ..., 5, 5, 5])
np.pad(a, (2, 3), 'linear_ramp', end_values=(5, -4))
array([ 5,  3,  1,  2,  3,  4,  5,  2, -1, -4])
np.pad(a, (2,), 'maximum')
array([5, 5, 1, 2, 3, 4, 5, 5, 5])
...
```

Padding is specified as such:
```
Number of values padded to the edges of each axis. ((before_1, after_1), ... (before_N, after_N)) unique pad widths for each axis. (before, after) or ((before, after),) yields same before and after pad for each axis. (pad,) or int is a shortcut for before = after = pad width for all axes.
```

Note that padding is specified across dimensions in the first -> last order, in contrast to last -> first as in PyTorch. Padding must be specified across all dimensions, and there is simple support for uniform padding.

2. (from [Twitter](https://x.com/_aidan_clark_/status/1923854993347707112?t=S-AaKFBbmKCpjoVj84jTgg&s=19)):

```python
torch.pad(t, dim=3, to=D)
torch.pad(t, shape=shape)
```

This provides more convenient padding specification for padding out to a given single dim shape / full size.

**New API wishlist:**
* Intuitive ordering (no more last -> first)
* Convenient / consistent padding for a single dim / subset of dims
* Complete set of padding modes matching real use cases
    * #57911 points out that PyTorch's circular padding mode only allows wrapping once, whereas numpy's wrap mode will wrap any number of times. This is an arbitrary restriction that requires a user to apply pad multiple times to get what they need
* Others ??? Please discuss below!

**Comment by soumith** [comment]
I like option (3), but atleast having a fill value that's "constant" (i.e. not just zero-padding)

**Comment by ngimel** [comment]
Tbh numpy API is even more confusing than current torch.nn.functional

**Comment by soumith** [comment]
thumbs up for Option 1

**Comment by soumith** [comment]
thumbs up for Option 2

**Comment by soumith** [comment]
thumbs up for Option 3

**Comment by ngimel** [comment]
2 is ambiguous for odd differences between current and desired shape, and doesn't allow asymmetric padding

**Comment by ngimel** [comment]
btw for option 3 we definitely want fill_value as @soumith suggested and also "mode" - reflect, circular, constant, replicate

**Comment by soumith** [comment]
I think implementing "circular" might be a bit too involved and the op might not see the light of the day; so I say we introduce circular at some later undertermined date

**Comment by ngimel** [comment]
We already have circular, it's just a question of calling existing op with a better API
